src/hotspot/share/runtime/os.cpp

```diff
@@ -1857,3 +1845,82 @@ void os::naked_sleep(jlong millis) {
   }
   naked_short_sleep(millis);
 }
+
+
+////// Implementation of pagesizeset_t
+
+// A pagesizeset_t is a set containing a set of page sizes.
+
+// sets the given page size
+void os::PagesizeSet::add(size_t pagesize) {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  _v |= pagesize;
+}
+
+// returns true if given page size is part of the set
+bool os::PagesizeSet::is_set(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  return _v & pagesize;
+}
+
+// returns the next smallest page size set in this set, or 0.
+size_t os::PagesizeSet::next_smaller(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  // mask out all pages sizes >= pagesize:
+  uintx v2 = _v & pagesize - 1;
+  if (v2 > 0) {
+    return round_down_power_of_2(v2);
+  }
+  return 0;
+}
+
+size_t os::PagesizeSet::next_larger(size_t pagesize) const {
+  assert(is_power_of_2(pagesize), "pagesize must be a power of 2: " INTPTR_FORMAT, pagesize);
+  if (pagesize == max_power_of_2<uintx>()) { // Shift by 32/64 would be UB
+    return 0;
+  }
+  int l = exact_log2(pagesize) + 1;
+  uintx v2 = _v >> l;
+  if (v2 == 0) {
+    return 0;
+  }
+  while ((v2 & 1) == 0) {
```

Comment:
Maybe the `count_trailing_zeros()` method could be used here instead of doing the same manually.